// components/FormInput/FormInput.js
import React, { memo } from 'react';
import { StyleSheet, TextInput, View } from 'react-native';
import PropTypes from 'prop-types';

const FormInput = memo(({
  value, label, onChangeText, ...props
}) => (
  <View style={styles.inputWrapper}>
    <TextInput
      {...props}
      style={styles.input}
      autoCorrect={false}
      disableFullscreenUI={true}
      autoCompleteType="off"
      underlineColorAndroid='transparent'
      placeholderTextColor='rgba(0, 0, 0, .3)'
      blurOnSubmit={false}
      value={value}
      placeholder={label}
      onChangeText={onChangeText}
    />
  </View>
));

FormInput.propTypes = {
  value: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  onChangeText: PropTypes.func.isRequired,
};

export default FormInput;

const styles = StyleSheet.create({
  input: {
    width: '100%',
    minHeight: 30,
    fontSize: 14,
    lineHeight: 16,
    padding: 0,
    margin: 0,
    borderBottomWidth: StyleSheet.hairlineWidth,
    borderColor: '#ccc',
  },
  inputWrapper: {
    marginVertical: 10,
  },
});
