// components/FlatButton/FlatButton.js
import React, { memo, useLayoutEffect } from 'react';
import {
  TouchableOpacity,
  Text,
  StyleSheet,
  ActivityIndicator,
  LayoutAnimation,
} from 'react-native';
import PropTypes from 'prop-types';

const FlatButton = memo(({ text = '', fetching = false, ...props }) => {
  const renderContent = () => {
    if (fetching) {
      return <ActivityIndicator color="#fff" size='small' />;
    }
    return <Text style={styles.text}>{text}</Text>;
  };

  useLayoutEffect(() => LayoutAnimation.easeInEaseOut());

  return (
    <TouchableOpacity
      style={[styles.btn, props.disabled && styles.disabled]}
      {...props}
    >
      {renderContent()}
    </TouchableOpacity>
  );
});

export default FlatButton;

FlatButton.propTypes = {
  text: PropTypes.string.isRequired,
  fetching: PropTypes.bool,
};

const styles = StyleSheet.create({
  btn: {
    paddingHorizontal: 20,
    paddingVertical: 8,
    backgroundColor: '#3e3ef9',
    borderRadius: 5,
    borderWidth: StyleSheet.hairlineWidth,
    borderColor: 'rgba(0, 0, 0, .1)',
    justifyContent: 'center',
    minWidth: 120,
    height: 38,
    shadowColor: '#3e3ef9',
    shadowOffset: {
      width: 0,
      height: 4,
    },
    shadowOpacity: 0.6,
    shadowRadius: 4,
    elevation: 2,
  },
  disabled: {
    backgroundColor: '#ccc',
    shadowColor: '#ccc',
  },
  text: {
    color: '#fff',
    fontSize: 16,
    lineHeight: 18,
    textAlign: 'center',
  },
});
