// navigation/RootNavigator.js
import { createStackNavigator, createAppContainer, createSwitchNavigator } from 'react-navigation';
import { useScreens } from 'react-native-screens';
import { fromBottom } from 'react-navigation-transitions';

import Main from 'screens/Main';
import LogIn from 'screens/LogIn';
import AuthLoadingScreen from 'screens/LoadingScreen';
import Modal from 'screens/Modal';
import { isIos } from 'utils';

useScreens();

const AuthNavigator = createStackNavigator({
  LogInScreen: LogIn,
});

const AppNavigator = createStackNavigator(
  {
    MainScreen: Main,
    ModalScreen: {
      screen: Modal,
    },
  },
  {
    initialRouteName: 'MainScreen',
    headerLayoutPreset: 'center',
    transitionConfig: () => fromBottom(400),
    mode: 'modal',
    cardStyle: {
      backgroundColor: isIos ? 'rgba(0, 0, 0, 0)' : 'rgba(0, 0, 0, .6)',
      opacity: 1.0,
    },
    transparentCard: true,
    defaultNavigationOptions: {
      headerStyle: {
        shadowColor: '#000',
        shadowOffset: {
          width: 0,
          height: 2,
        },
        shadowOpacity: 0.15,
        shadowRadius: 6,
        elevation: 4,
        borderBottomColor: 'rgba(0, 0, 0, .1)',
        height: 47,
      },
    },
  },
);

const Navigation = createAppContainer(
  createSwitchNavigator(
    {
      Auth: AuthNavigator,
      App: AppNavigator,
      AuthLoading: AuthLoadingScreen,
    },
    {
      initialRouteName: 'AuthLoading',
    },
  ),
);

export default Navigation;
