// components/Logo/Logo.js
import React, { memo } from 'react';
import { View, StyleSheet, Text } from 'react-native';
import AntDesign from 'react-native-vector-icons/AntDesign';
import * as Animatable from 'react-native-animatable';
import PropTypes from 'prop-types';

const Logo = memo(
  ({
    text = '', size = 30, color = '#222', animated = false,
  }) => (
    <View style={styles.wrapper}>
      <Animatable.View
        useNativeDriver
        animation={animated ? 'zoomIn' : null}
        delay={200}
      >
        <AntDesign name="github" size={size} color={color} />
      </Animatable.View>

      <Animatable.View
        useNativeDriver
        animation={animated ? 'fadeInRight' : null}
      >
        <Text style={[styles.title, { color, fontSize: size * 0.65 }]}>
          {text}
        </Text>
      </Animatable.View>
    </View>
  ),
);

const styles = StyleSheet.create({
  wrapper: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  title: {
    fontSize: 18,
    fontWeight: '600',
    marginLeft: 8,
  },
});

Logo.propTypes = {
  text: PropTypes.string.isRequired,
  size: PropTypes.number,
  color: PropTypes.string,
  animated: PropTypes.bool,
};

export default Logo;
