// components/Up/Up.js
import React, { memo } from 'react';
import { TouchableOpacity, StyleSheet } from 'react-native';
import PropTypes from 'prop-types';
import Icon from 'react-native-vector-icons/Feather';
import * as Animatable from 'react-native-animatable';

const handlePress = (element) => {
  if (element) {
    element.current.scrollToOffset({ offset: 0 });
  }
};

const Btn = Animatable.createAnimatableComponent(TouchableOpacity);

// eslint-disable-next-line react/display-name
const Up = memo(({
  visible = false, size = 20, style = {}, target,
}) => (visible ? (
    <Btn
      useNativeDriver
      animation="fadeInUpBig"
      style={[styles.wrapper, { borderRadius: size / 2 }, style]}
      onPress={() => handlePress(target)}
      activeOpacity={0.75}
    >
      <Icon name="chevrons-up" size={size} color="#444" />
    </Btn>
) : null));

Up.propTypes = {
  target: PropTypes.object.isRequired,
  visible: PropTypes.bool.isRequired,
  size: PropTypes.string,
  style: PropTypes.object,
};

export default Up;

const styles = StyleSheet.create({
  wrapper: {
    padding: 10,
    backgroundColor: 'rgba(255, 255, 255, .95)',
    justifyContent: 'center',
    alignItems: 'center',
    position: 'absolute',
    borderWidth: StyleSheet.hairlineWidth,
    borderColor: 'rgba(0, 0, 0, .1)',
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.15,
    shadowRadius: 6,
    elevation: 3,
  },
});
