// components/Preloader/Preloader.js
import React from 'react';
import {
  ActivityIndicator, StyleSheet, Text, View,
} from 'react-native';
import PropTypes from 'prop-types';

export const Preloader = ({ color = '#007AFF', absolute = false, visible = false }) => (visible ? (
    <View style={[styles.wrapper, absolute && styles.absolute]}>
      <View style={styles.innerWrapper}>
        <ActivityIndicator size='large' color={color} />
        <Text style={styles.text}>Loading...</Text>
      </View>
    </View>
) : null);

const styles = StyleSheet.create({
  wrapper: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'rgb(244, 244, 244)',
  },
  absolute: {
    ...StyleSheet.absoluteFillObject,
    backgroundColor: 'rgba(244, 244, 244, .8)',
  },
  text: {
    marginTop: 8,
    color: '#888',
  },
  innerWrapper: {
    padding: 15,
    backgroundColor: '#fff',
    borderWidth: StyleSheet.hairlineWidth,
    borderColor: 'rgba(0, 0, 0, .1)',
    borderRadius: 10,
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.1,
    shadowRadius: 6,
    elevation: 5,
  },
});

Preloader.propTypes = {
  color: PropTypes.string,
  absolute: PropTypes.bool,
  visible: PropTypes.bool.isRequired,
};

export default Preloader;
